lib/common/busypoll.c

```diff
@@ -0,0 +1,444 @@
+                    h2o_perror("inet_ntop couldn't convert addr");
+                }
+
+                fprintf(stderr,
+                        "thread: %lu claimed NIC %s [%d]. NAPI IDs MISmatch [%d from %s], is_local: %d, laddr: %s, lport: %d, "
+                        "family: %d\n",
+                        thread_index, nic_claimed, napi_id_claimed, napi_id, mismatch_nic_name, verdict, laddr,
+                        ntohs(localaddr.sin_port), localaddr.sin_family);
+                return;
+            }
+        } else {
+            /* if this thread claimed the loopback device, let's make sure the incoming
+             * connection is actually a loopback connection with napi id of 0
+             */
+            if (napi_id == 0) {
+                if (verdict == 1) {
+                    /* is_local returned 1, as expected, so this socket is fine */
+                    return;
+                }
+
+                if (verdict == 0) {
+                    /* is_local returned 0, so this NAPI ID matches and is correct but is not a local connection */
+                    fprintf(stderr, "thread: %lu claimed loopback, got a loopback connection, but conn is NOT local\n",
+                            thread_index);
+                }
+            } else {
+                /* in this case the the connection has a mismatched NAPI ID. Let's check if its local to help debug */
+                fprintf(stderr, "thread: %lu claimed loopback [%d] but received non-zero NAPI ID [%d] is local: %d\n", thread_index,
+                        napi_id_claimed, napi_id, verdict);
+                return;
+            }
+        }
+    } else {
+        /* a CPU has not been claimed yet, so claim one */
+        int listener_fd = h2o_socket_get_fd(listener);
+        struct ifreq ifr = {};
+
+        if (get_nic_name_by_napi(sockfd, napi_id, &ifr) == -1) {
+            fprintf(stderr, "SIOCGIFNAME_BY_NAPI_ID failed ? %d\n", napi_id);
+            h2o_perror("SIOCGIFNAME_BY_NAPI_ID failed");
+        }
+
+        fprintf(stderr, "thread %zd NAPI id: %u, iface: %s, listener fd: %d\n", thread_index, napi_id, ifr.ifr_name, listener_fd);
+
+        int found = assign_nic_map_cpu(ifr.ifr_name, thread_index, napi_id, listener_fd, 0, nic_to_cpu_map, nic_count);
+
+        if (!found) {
+            fprintf(stderr, "no CPU found for thread with NAPI from NIC: %s:%d\n", ifr.ifr_name, napi_id);
+            fprintf(stderr, "busy-poll-cpu-map configuration or queue setup is buggy -- exiting h2o now.\n");
+            exit(-1);
+        }
+    }
+
+    return;
+}
+
+void h2o_busypoll_bind_interface(int fd, const char *iface)
+{
+    bind_interface(fd, iface);
+}
+
+void h2o_busypoll_set_opts(uint32_t ifindex, uint32_t defer_hard_irqs, uint64_t gro_flush_timeout, uint64_t irq_suspend_timeout)
+{
+    setup_queue(ifindex, defer_hard_irqs, gro_flush_timeout, irq_suspend_timeout);
```

Comment:
this should probably either iterate across the NAPI IDs for an interface setting the options, or `setup_queue` internally should iterate across a range of queue numbers setting these settings

Reply:
thank you, i opted for the latter approach in 29542d8d3383aba2a4c59334be6de0f3c871202b. `setup_queue` iterates internally and saves the napi ids for later use.